### PR TITLE
fix($vdom): Don't replace input for text-like type change, fix #6313

### LINKED ISSUE #6313: conditional change of <input> type attribute not working.
### Version
2.4.2

### Reproduction link
[https://jsfiddle.net/trusktr/50wL7mdz/52198/](https://jsfiddle.net/trusktr/50wL7mdz/52198/)

### Steps to reproduce
Try to type in the text field. It will keep clearing your input every second.

### What is expected?
It should toggle between showing and hiding the password.

### What is actually happening?
It is destroying the input (or clearing its value) each time the value of `hidden` changes.

### PR DESCRIPTION
**Pull request by nickmessing**
**What kind of change does this PR introduce?** (check at least one)

- [x] Bugfix
- [ ] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch for v2.x (or to a previous version branch), _not_ the `master` branch
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] All tests are passing: https://github.com/vuejs/vue/blob/dev/.github/CONTRIBUTING.md#development-setup
- [x] New/updated tests are included

If adding a **new feature**, the PR's description includes:
- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)
 
**Other information:**
fix #6313

### CODE REVIEW

File: src/core/vdom/patch.js [hunk 1]
```diff
@@ -48,14 +48,12 @@ function sameVnode (a, b) {
   )
 }
 
-// Some browsers do not support dynamically changing type for <input>
-// so they need to be treated as different nodes
```
Comment (nickmessing) [inline src/core/vdom/patch.js]: This is not relevant since it affects browsers that Vue doesn't support.

**Comment by nickmessing** [comment]
I suggest considering merging together with #6227

File: src/core/vdom/patch.js [hunk 2]
```diff
@@ -27,6 +25,8 @@ import {
   isPrimitive
 } from '../util/index'
 
+const isTextInputType = makeMap(TEXT_INPUT_TYPES)
```
Comment (javoski) [inline src/core/vdom/patch.js]: Maybe we could move function `isTextInputType` to [src/platforms/web/util/element.js](https://github.com/vuejs/vue/blob/dev/src/platforms/web/util/element.js).

File: test/unit/modules/vdom/patch/edge-cases.spec.js [hunk 3]
```diff
@@ -135,4 +135,27 @@ describe('vdom patch: edge cases', () => {
...
+      data: { show: false },
+      template: `
+        <div>
+          <input :type="show ? 'text' : 'password'">
+        </div>
+      `
+    }).$mount()
+    const node = vm.$el.children[0]
+    vm.$el.children[0].value = 'test'
+    vm.ok = true
```
Comment (nickmessing) [inline test/unit/modules/vdom/patch/edge-cases.spec.js]: :laughing:, lol, that is a dumb mistake :D

File: test/unit/modules/vdom/patch/edge-cases.spec.js [hunk 4]
```diff
@@ -135,4 +135,27 @@ describe('vdom patch: edge cases', () => {
...
+        </div>
+      `
+    }).$mount()
+    const node = vm.$el.children[0]
+    vm.$el.children[0].value = 'test'
+    vm.show = true
+    waitForUpdate(() => {
+      expect(vm.$el.children[0]).toBe(node)
+      expect(vm.$el.children[0].value).toBe('test')
+      vm.show = false
```
Comment (dsonet) [inline test/unit/modules/vdom/patch/edge-cases.spec.js]: Should also assert the input type attribute.

File: src/core/vdom/patch.js (hunk 2 above)
Comment (nickmessing) [inline src/core/vdom/patch.js]: @javoski, done, any more feedback?

File: test/unit/modules/vdom/patch/edge-cases.spec.js (hunk 4 above)
Comment (nickmessing) [inline test/unit/modules/vdom/patch/edge-cases.spec.js]: @dsonet, done, any more feedback?

**Comment by HerringtonDarkholme** [comment]
I would suggest to apply this change to vdom's patching. So users can use fully dynamic expression. For now it seems to only handle ternary expression.

**Comment by yyx990803** [comment]
@nickmessing I removed the ternary support parts to make this PR more focused. I have some ideas on how to fully support dynamic types on `v-model` but that will be part of 2.5.
